src/Razor/src/Microsoft.VisualStudio.LanguageServerClient.Razor/DefaultRazorLanguageServerCustomMessageTarget.cs

```diff
@@ -651,77 +625,79 @@ public override async Task<VSInternalWrapWithTagResponse> RazorWrapWithTagAsync(
                     {
-                        var csharpRequestParams = new FoldingRangeParams()
-                        {
-                            TextDocument = new()
-                            {
-                                Uri = csharpDocument.Uri
-                            }
-                        };
-
-                        var request = await _requestInvoker.ReinvokeRequestOnServerAsync<FoldingRangeParams, IEnumerable<FoldingRange>?>(
-                            Methods.TextDocumentFoldingRange.Name,
-                            RazorLSPConstants.RazorCSharpLanguageServerName,
-                            SupportsFoldingRange,
-                            csharpRequestParams,
-                            cancellationToken).ConfigureAwait(false);
-
-                        var result = request.Result;
-                        if (result is not null)
+                        TextDocument = new()
                         {
-                            csharpRanges.AddRange(result);
+                            Uri = csharpSnapshot.Uri
                         }
+                    };
+
+                    var request = await _requestInvoker.ReinvokeRequestOnServerAsync<FoldingRangeParams, IEnumerable<FoldingRange>?>(
+                        csharpSnapshot.Snapshot.TextBuffer,
+                        Methods.TextDocumentFoldingRange.Name,
+                        RazorLSPConstants.RazorCSharpLanguageServerName,
+                        SupportsFoldingRange,
+                        csharpRequestParams,
+                        cancellationToken).ConfigureAwait(false);
+
+                    var result = request?.Response;
+                    if (result is null)
+                    {
+                        csharpRanges = null;
                     }
-                }, cancellationToken);
-
-            }
+                    else
+                    {
+                        csharpRanges.AddRange(result);
+                    }
+                }
+            }, cancellationToken);
 
-            var htmlDocument = GetHtmlDocumentSnapshsot(foldingRangeParams.TextDocument.Uri);
             var htmlRanges = new List<FoldingRange>();
             var htmlTask = Task.CompletedTask;
-            if (htmlDocument is not null)
+            htmlTask = Task.Run(async () =>
             {
-                htmlTask = Task.Run(async () =>
-                {
-                    var synchronized = await _documentSynchronizer.TrySynchronizeVirtualDocumentAsync(
-                        foldingRangeParams.HostDocumentVersion, htmlDocument, cancellationToken);
+                var htmlDocumentUri = GetHtmlFilePath(foldingRangeParams.TextDocument.Uri);
+                var (synchronized, htmlDocument) = await _documentSynchronizer.TrySynchronizeVirtualDocumentAsync<HtmlVirtualDocumentSnapshot>(
+                    foldingRangeParams.HostDocumentVersion, foldingRangeParams.TextDocument.Uri, htmlDocumentUri, cancellationToken);
```

Comment:
This whole review I've been thinking "its weird that you can only get one virtual document. What if you want both?", and now I have my answer :)

Is it worth using the old overloads for this scenario? There is some amount of extra complexity here with locks and async machinery that is essentially unnecessary. We only need to synchronize once, then get both snapshots. Maybe if instead of `Obsolete`ing, we renamed the old methods to `WaitForDocumentSynchronize` or something, it would be less likely to make people think they actually _do_ anything to the snapshots?

Reply:
I guess, upon hitting the bottom of the file, this is the only case where we actually use both snapshots. The more common pattern is to use just one, so maybe we don't worry about this?

Reply:
Yeah, we don't use both in very many places, and they're also synchronizing different things. ie HTML virtual documents being synchronized does not mean that CSharp necessarily is. This is also how things were before (synchronize a single virtualDocument) so I'd say we should make the smaller change since I already hit 50+ files.

Reply:
> HTML virtual documents being synchronized does not mean that CSharp necessarily is.

Interesting. Certainly not the impression the old API gave, which means that this new API is objectively better, and possibly we were masking some other bugs. So yeah, lets keep it.